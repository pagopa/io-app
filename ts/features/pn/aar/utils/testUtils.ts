import { FiscalCode, NonEmptyString } from "@pagopa/ts-commons/lib/strings";
import { MessageBodyMarkdown } from "../../../../../definitions/backend/MessageBodyMarkdown";
import { MessageSubject } from "../../../../../definitions/backend/MessageSubject";
import { ThirdPartyMessage as PnThirdPartyMessage } from "../../../../../definitions/pn/ThirdPartyMessage";
import { EphemeralAarMessageDataActionPayload } from "../store/actions";
import { AARFlowState, sendAARFlowStates } from "./stateUtils";

export const sendAarMockStateFactory: Record<
  AARFlowState["type"],
  () => AARFlowState
> = {
  none: () => ({ type: "none" }),
  displayingAARToS: () => ({
    type: "displayingAARToS",
    qrCode: "https://www.google.com"
  }),
  fetchingQRData: () => ({
    type: "fetchingQRData",
    qrCode: "https://www.google.com"
  }),
  fetchingNotificationData: () => ({
    type: "fetchingNotificationData",
    iun: "000000000001",
    recipientInfo: {
      denomination: "Mario Rossi",
      taxId: "RSSMRA74D22A001Q"
    }
  }),
  displayingNotificationData: () => ({
    type: "displayingNotificationData",
    recipientInfo: {
      denomination: "Mario Rossi",
      taxId: "RSSMRA74D22A001Q"
    },
    notification: {},
    iun: "000000000001",
    pnServiceId: "SERVICEID123" as NonEmptyString,
    mandateId: "MANDATE123"
  }),
  notAddresseeFinal: () => ({
    type: "notAddresseeFinal",
    recipientInfo: {
      denomination: "Mario Rossi",
      taxId: "RSSMRA74D22A001Q"
    },
    qrCode: "https://www.google.com",
    iun: "000000000001"
  }),
  ko: () => ({
    type: "ko",
    previousState: { type: "none" },
    debugData: {
      phase: "Entry Point",
      reason: "Sample reason"
    }
  })
};

export const sendAarStateNames = Object.values(sendAARFlowStates);
export const sendAarMockStates = sendAarStateNames.map(t =>
  sendAarMockStateFactory[t]()
);
export const mockEphemeralAarMessageDataActionPayload: EphemeralAarMessageDataActionPayload =
  {
    iun: "IUN123" as NonEmptyString,
    thirdPartyMessage: { foo: "bar" } as PnThirdPartyMessage,
    fiscalCode: "TAXCODE123" as FiscalCode,
    pnServiceID: "SERVICEID123" as NonEmptyString,
    markdown: "*".repeat(81) as MessageBodyMarkdown,
    subject: "subject" as MessageSubject,
    mandateId: "MANDATE123"
  };
