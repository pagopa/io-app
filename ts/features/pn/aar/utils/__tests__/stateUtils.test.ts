import {
  aarErrors,
  isAarErrorRetriable,
  isValidAARStateTransition,
  sendAARFlowStates,
  testable,
  validAARStatusTransitions
} from "../stateUtils";
import { sendAarStateNames } from "../testUtils";

const { finalErrors, retriableErrors } = testable!;

describe("stateUtils", () => {
  describe("isValidAARStateTransition function", () => {
    sendAarStateNames.forEach(currentType => {
      sendAarStateNames.forEach(nextType => {
        const allowedNextStates = validAARStatusTransitions.get(currentType);
        const isAllowed = allowedNextStates?.has(nextType) ?? false;
        it(
          isAllowed
            ? `Should allow a valid from '${currentType}' to '${nextType}'`
            : `Should reject an invalid from '${currentType}' to '${nextType}'`,
          () => {
            expect(isValidAARStateTransition(currentType, nextType)).toBe(
              isAllowed
            );
          }
        );
      });
    });
  });
  describe("isAarErrorRetriable", () => {
    Object.values(retriableErrors).forEach(value => {
      it(`should return true for retriable error: ${value}`, () => {
        const error = aarErrors[value];
        expect(isAarErrorRetriable(error)).toBe(true);
      });
    });

    Object.values(finalErrors).forEach(value => {
      it(`should return false for final (non-retriable) error: ${value}`, () => {
        const error = aarErrors[value];
        expect(isAarErrorRetriable(error)).toBe(false);
      });
    });
  });
  describe("snapshots", () => {
    it("validTransitions", () => {
      expect(validAARStatusTransitions).toMatchSnapshot();
    });
    it("flowStates", () => {
      expect(sendAARFlowStates).toMatchSnapshot();
    });
    it("aarErrors", () => {
      expect(aarErrors).toMatchSnapshot();
    });
  });
});
