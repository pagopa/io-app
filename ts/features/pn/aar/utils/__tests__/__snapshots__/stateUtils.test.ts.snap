// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stateUtils snapshots flowStates 1`] = `
{
  "displayingAARToS": "displayingAARToS",
  "displayingNotificationData": "displayingNotificationData",
  "fetchingNotificationData": "fetchingNotificationData",
  "fetchingQRData": "fetchingQRData",
  "ko": "ko",
  "none": "none",
  "notAddresseeFinal": "notAddresseeFinal",
}
`;

exports[`stateUtils snapshots validTransitions 1`] = `
Map {
  "none" => Set {
    "displayingAARToS",
  },
  "displayingAARToS" => Set {
    "fetchingQRData",
  },
  "fetchingQRData" => Set {
    "fetchingNotificationData",
    "notAddresseeFinal",
    "ko",
  },
  "fetchingNotificationData" => Set {
    "displayingNotificationData",
    "ko",
  },
  "displayingNotificationData" => Set {},
  "notAddresseeFinal" => Set {},
  "ko" => Set {
    "fetchingQRData",
    "fetchingNotificationData",
  },
}
`;
